botcore/utils/messages.py

```diff
@@ -0,0 +1,306 @@
+import asyncio
+import random
+import re
+from functools import partial
+from io import BytesIO
+from typing import Callable, Sequence
+
+import discord
+from discord.ext import commands
+
+from botcore.utils import scheduling
+from botcore.utils.logging import get_logger
+
+
+log = get_logger(__name__)
+
+
+def reaction_check(
+    reaction: discord.Reaction,
+    user: discord.abc.User,
+    *,
+    message_id: int,
+    allowed_emoji: Sequence[str],
+    allowed_users: Sequence[int],
+    mod_roles: Sequence[int],
+    allow_mods: bool = True,
+) -> bool:
+    """
+    Check if a reaction's emoji and author are allowed and the message is `message_id`.
+
+    If the user is not allowed, remove the reaction. Ignore reactions made by the bot.
+    If `allow_mods` is True, allow users with `mod_roles` even if they're not in `allowed_users`.
+    """
+    right_reaction = (
+        not user.bot
+        and reaction.message.id == message_id
+        and str(reaction.emoji) in allowed_emoji
+    )
+    if not right_reaction:
+        return False
+
+    is_moderator = (
+        allow_mods
+        and any(role.id in mod_roles for role in getattr(user, "roles", []))
+    )
+
+    if user.id in allowed_users or is_moderator:
+        log.trace(f"Allowed reaction {reaction} by {user} on {reaction.message.id}.")
+        return True
+    else:
+        log.trace(f"Removing reaction {reaction} by {user} on {reaction.message.id}: disallowed user.")
+        scheduling.create_task(
+            reaction.message.remove_reaction(reaction.emoji, user),
+            suppressed_exceptions=(discord.HTTPException,),
+            name=f"remove_reaction-{reaction}-{reaction.message.id}-{user}"
+        )
+        return False
+
+
+async def wait_for_deletion(
+    bot: commands.Bot,
+    message: discord.Message,
+    user_ids: Sequence[int],
```

Comment:
```suggestion
    allowed_users: Sequence[int],
```
For a consistent interface across botcore.